### PR TITLE
v1.4: Support non_null connection node types

### PR DESCRIPTION
**Pull request by bruce**
This PR adds the ability to mark a connection node type as non-nullable, supporting #107—for v1.4 (stable), at least.

This PR also:

- Removes the JSON codec (no longer needed for development of this package)
- Adds some minor formatting rules, relaxing parens for `connection`
- Modernizes the Travis CI config somewhat

## Examples

```elixir
# Standard connection
connection node_type: non_null(:team)

# Connection, customizing edge
connection node_type: non_null(:user) do
   edge do
     field :role, :string
   end
end

# Named connection
connection :favorite_pets, node_type: non_null(:pet)
```

Don't try to do `list_of`, of course ;-) (The spec specifically says list node types aren't supported, and I am very thankful. How strange that would be...)

### CODE REVIEW

**Comment by jeroenvisser101** [comment]
@bruce I might be mistaken, but this only allows users to specify a non_null type on the list elements, but not on the list itself (for instance, when the list can be empty or not but never `nil`), right?

File: lib/absinthe/relay/connection/notation.ex [hunk 1]
```diff
@@ -210,17 +250,34 @@ defmodule Absinthe.Relay.Connection.Notation do
...
     edge_type = naming.edge_type_identifier
     [private_category_node, private_key_node] = @private_node_type_identifier_path
     [private_category_base, private_key_base] = @private_base_identifier_path
 
+    [private_category_non_null_node_type, private_key_non_null_node_type] =
+      @private_non_null_node_type_path
+
     quote do
       field :page_info, type: non_null(:page_info)
       field :edges, type: list_of(unquote(edge_type))
```
Comment (jeroenvisser101) [inline lib/absinthe/relay/connection/notation.ex]: So this type would preferably also be configurable to be `non_null(list_of(...))`

**Comment by jgwmaxwell** [comment]
Is there anything that I can offer to help expedite merging this? It's a huge issue with generated types and would be great to see resolved.

**Comment by harrygr** [comment]
I too would love to see this merged.

**Comment by onx2** [comment]
Is this able to be merged? It would be a huge help if we didn't have to do multiple null checks in TS when working with generated types.

**Comment by benwilson512** [comment]
@onx2 there are no plans to merge this in for 1.4, in fact I'll be closing it. However this PR for 1.5 is in good shape https://github.com/absinthe-graphql/absinthe_relay/pull/158 so I'll see if I can get it merged.

**Pull request closed without merging**